library/aes.c

```diff
@@ -1360,34 +1360,21 @@ int mbedtls_aes_crypt_cfb128(mbedtls_aes_context *ctx,
         return MBEDTLS_ERR_AES_BAD_INPUT_DATA;
     }
 
-    if (mode == MBEDTLS_AES_DECRYPT) {
-        while (length--) {
-            if (n == 0) {
-                ret = mbedtls_aes_crypt_ecb(ctx, MBEDTLS_AES_ENCRYPT, iv, iv);
-                if (ret != 0) {
-                    goto exit;
-                }
+    while (length--) {
+        if (n == 0) {
+            ret = mbedtls_aes_crypt_ecb(ctx, MBEDTLS_AES_ENCRYPT, iv, iv);
+            if (ret != 0) {
+                goto exit;
             }
-
-            c = *input++;
-            *output++ = (unsigned char) (c ^ iv[n]);
-            iv[n] = (unsigned char) c;
-
-            n = (n + 1) & 0x0F;
         }
-    } else {
-        while (length--) {
-            if (n == 0) {
-                ret = mbedtls_aes_crypt_ecb(ctx, MBEDTLS_AES_ENCRYPT, iv, iv);
-                if (ret != 0) {
-                    goto exit;
-                }
-            }
-
-            iv[n] = *output++ = (unsigned char) (iv[n] ^ *input++);
 
-            n = (n + 1) & 0x0F;
+        c = *input;
+        iv[n] = *output++ = *input++ ^ iv[n];
+        if (mode == MBEDTLS_AES_DECRYPT) {
```

Comment:
So it's pretty horrible that this function processes data one byte at a time, but it looks like the separation into almost-identical encrypt/decrypt loops will have been done for performance.

With this change, we now have an `if` (potential branch) within a tight loop - do we have any idea of the impact on performance?

Reply:
If you’ve got an update benchmark, could you raise a PR for that bit?